test/EthLightClient.js

```diff
@@ -0,0 +1,150 @@
+const { expect } = require("chai");
+const { ethers, upgrades } = require("hardhat");
+const block1 = require('./proof1.json');
+const block2 = require('./proof2.json');
+
+function hexToBytes(hex) {
+    for (var bytes = [], c = 0; c < hex.length; c += 2)
+        bytes.push(parseInt(hex.substr(c, 2), 16));
+    return bytes;
+}
+
+const DEBUG = true;
+
+function debug(msg){
+    if(DEBUG) console.log(msg);
+}
+
+
+describe("Ethereum Light Client", function () {
+
+    let ELC, ELCDeployer;
+
+    beforeEach(async function () {
+        const [owner] = await ethers.getSigners();
+
+        ELCDeployer = await ethers.getContractFactory("TesterEthereumLightClient");
+
+        rlpHeader = block1.header_rlp.substring(2);
+
+        bytes = hexToBytes(rlpHeader);
+
+        ELC = await upgrades.deployProxy(ELCDeployer, [bytes]);
+
+        await ELC.deployed();
+    });
+
+    it("Dummy Add Block TELC", async function () {
+
+        hashes = ['68776803263250831790672116253057717414537012912849465124179566887710318158907', '25', '12', '9999999', '1543254325', '2154326543656436543653']
+
+        console.log(ELC.address);
+
+        console.log((await ELC.canonicalHead()).toString())
+
+        await ELC.dummmyAddBlockHeader(hashes[0], 400 ,hashes[1]);
+
+        CHead = (await ELC.canonicalHead()).toString();
+
+        parentHash = (await ELC.blocks(hashes[1])).parentHash.toString();
+
+        expect(parentHash).to.equal(hashes[0]);
+
+        expect(CHead).to.equal(hashes[1]);
+
+
+
+        await ELC.dummmyAddBlockHeader(hashes[1], 250, hashes[2]);
+
+        CHead = (await ELC.canonicalHead()).toString();
+
+        parentHash = (await ELC.blocks(hashes[2])).parentHash.toString();
+
+        expect(parentHash).to.equal(hashes[1]);
+
+        expect(CHead).to.equal(hashes[2]);
+    });
+
+    describe("Forking tests", function() {
```

Comment:
Is there some way we can test the real `addBlockHeader` rather than the dummy one? There are lots of logics are in the real one that are omitted in the dummy test client

Reply:
The full canonical chain replacement test could and should be run using real blocks. The forking test however, I was having the problem of finding real world data of a Ropsten fork that would be needed for testing. I will work on appending a chain replacement test using real ropsten blocks and the real addBlockHeader today. Let me know if anyone has an idea for how to get test block data that can be used in a forking test. I had considered using the ganache functionality to fork an existing testnet and pulling blocks from that, however I havent followed up with that idea in a bit.

Reply:
Actually, since block numbers have to overlap to test a chain replacement, its not actually just straightforward to use ropsten. Maybe look at the same blocks on ropsten/mainnet or other testnet for that one could be possible. However yeah, this is sort of the issue I ran into where it was seeming that it would be very hard to obtain proper test data from a live environment for the conditions I needed to test

Reply:
Ganashe has some functionality to fork live chains, I will look into using this

Reply:
Seems ganache local fork does not compute hashes correctly, at least the cli tool throws "ethash local wrong!". I think that we may run into problems using ganache in general for this problem.

Reply:
Something like this would probably work, however I've never really set up Geth nodes before so it would be pretty involved. 
https://ethereum.stackexchange.com/questions/2851/how-can-i-reliably-induce-a-blockchain-fork-for-testing-purposes

Reply:
Also it seems the hardhat built in local networks are different than ganache, so I may try using the hardhat fork functionality as well 🤞